src/docfx/build/context/Input.cs

```diff
@@ -0,0 +1,115 @@
+
+using System;
+using System.Collections.Concurrent;
+using System.IO;
+
+namespace Microsoft.Docs.Build
+{
+    internal class Input
+    {
+        private readonly string _docsetPath;
+        private readonly string _fallbackPath;
+        private readonly Config _config;
+        private readonly RestoreGitMap _restoreMap;
+        private readonly ConcurrentDictionary<FilePath, byte[]> _gitBlobCache = new ConcurrentDictionary<FilePath, byte[]>();
+
+        public Input(string docsetPath, string fallbackPath, Config config, RestoreGitMap restoreMap)
+        {
+            _config = config;
+            _restoreMap = restoreMap;
+            _docsetPath = Path.GetFullPath(docsetPath);
+            _fallbackPath = fallbackPath is null ? null : Path.GetFullPath(fallbackPath);
+        }
+
+        public bool Exists(FilePath file)
+        {
+            var (basePath, path, commit) = ResolveFilePath(file);
+
+            if (basePath is null)
+            {
+                return false;
+            }
+
+            if (commit is null)
+            {
+                return File.Exists(Path.Combine(basePath, path));
+            }
+
+            throw new NotSupportedException("Checking if a file exists in a git repo");
+        }
+
+        public bool TryGetPhysicalPath(FilePath file, out string physicalPath)
+        {
+            var (basePath, path, commit) = ResolveFilePath(file);
+
+            if (basePath != null && commit is null)
+            {
+                var fullPath = Path.Combine(basePath, path);
+                if (File.Exists(fullPath))
+                {
+                    physicalPath = fullPath;
+                    return true;
+                }
+            }
+
+            physicalPath = null;
+            return false;
+        }
+
+        public string ReadString(FilePath file)
+        {
+            using (var reader = ReadText(file))
+            {
+                return reader.ReadToEnd();
+            }
+        }
+
+        public TextReader ReadText(FilePath file)
+        {
+            var (basePath, path, commit) = ResolveFilePath(file);
+
+            if (basePath is null)
+            {
+                throw new NotSupportedException($"Cannot read file path '{file}'");
+            }
+
+            if (commit is null)
+            {
+                return File.OpenText(Path.Combine(basePath, path));
+            }
+
+            var bytes = _gitBlobCache.GetOrAdd(file, aFile => GitUtility.ReadBytes(_fallbackPath, aFile.Path, aFile.Commit))
+                ?? throw new InvalidOperationException($"Error reading '{file}'");
+
+            return new StreamReader(new MemoryStream(bytes, writable: false));
+        }
+
+        private (string basePath, string path, string commit) ResolveFilePath(FilePath file)
+        {
+            switch (file.Origin)
+            {
+                case FileOrigin.Default when file.Commit is null:
```

Comment:
>case [](start = 16, length = 4)

From input abstraction view of point, it's better avoid to invloving too many business logic in Input.

Default, Dependency, Fallback, Template can all map to Physical, Git and Memory.

Considering that in the future we add new business dependency model, we don't want to change Input to support it.

Reply:
The design goal of `FilePath` is serializable and roamable (machine independent), thus it is designed this way. Using __physical/git/memory__ is machine dependent.

Reply:
physical/git/memory is not always be machine dependent, and also, we should go details about the puporse of serializable/roadmable.

---
In reply to: [320084723](https://github.com/dotnet/docfx/pull/5034#discussion_r320084723) [](ancestors = 320084723)